File: python/x402/src/x402_tron/mechanisms/facilitator/gasfree.py
"""
GasFreeFacilitatorMechanism - GasFree payment scheme facilitator mechanism for TRON.
"""

import logging
from typing import Any, Optional

from x402_tron.address import AddressConverter, TronAddressConverter
from x402_tron.config import NetworkConfig
from x402_tron.mechanisms.facilitator.base_exact import BaseExactFacilitatorMechanism
from x402_tron.types import PaymentPermit, PaymentRequirements
from x402_tron.utils.gasfree import (
    GASFREE_PERMIT_TRANSFER_TYPES,
    GasFreeAPIClient,
    get_gasfree_domain,
)


class GasFreeFacilitatorMechanism(BaseExactFacilitatorMechanism):
    """GasFree facilitator mechanism for TRON"""

    def scheme(self) -> str:
        return "gasfree_exact"

    def _get_address_converter(self) -> AddressConverter:
        return TronAddressConverter()

    async def _verify_signature(
        self,
        permit: PaymentPermit,
        signature: str,
        network: str,
    ) -> bool:
        """Verify GasFree TIP-712 signature"""
        controller = NetworkConfig.get_gasfree_controller_address(network)
        chain_id = NetworkConfig.get_chain_id(network)
        converter = self._address_converter

        # Reconstruct GasFree message from the mapped permit
        message = {
            "token": converter.to_evm_format(permit.payment.pay_token),
            "serviceProvider": converter.to_evm_format(permit.fee.fee_to),
            "user": converter.to_evm_format(permit.buyer),
            "receiver": converter.to_evm_format(permit.payment.pay_to),
            "value": int(permit.payment.pay_amount),
            "maxFee": int(permit.fee.fee_amount),
            "deadline": int(permit.meta.valid_before),
            "version": 1,
            "nonce": int(permit.meta.nonce),
        }

        domain = get_gasfree_domain(chain_id, controller)

        self._logger.info(f"[VERIFY GASFREE] Domain: {domain}")
        self._logger.info(f"[VERIFY GASFREE] Message: {message}")

        return await self._signer.verify_typed_data(
            address=permit.buyer,
            domain=domain,
            types=GASFREE_PERMIT_TRANSFER_TYPES,
            message=message,
            signature=signature,
        )

    async def _settle_payment_only(
        self,
        permit: PaymentPermit,
        signature: str,
        requirements: PaymentRequirements,
    ) -> Optional[str]:
        """Settle GasFree transaction via official HTTP API"""
        network = requirements.network
        controller = NetworkConfig.get_gasfree_controller_address(network)
        chain_id = NetworkConfig.get_chain_id(network)
        api_base_url = NetworkConfig.get_gasfree_api_base_url(network)
        api_key = NetworkConfig.get_gasfree_api_key(network)
        api_secret = NetworkConfig.get_gasfree_api_secret(network)
        api_client = GasFreeAPIClient(api_base_url, api_key, api_secret)
        converter = self._address_converter

        # Build the message for the API
        message = {
            "token": converter.to_evm_format(permit.payment.pay_token),
            "serviceProvider": converter.to_evm_format(permit.fee.fee_to),
            "user": converter.to_evm_format(permit.buyer),
            "receiver": converter.to_evm_format(permit.payment.pay_to),
            "value": str(permit.payment.pay_amount),
            "maxFee": str(permit.fee.fee_amount),
            "deadline": str(permit.meta.valid_before),
            "version": 1,
            "nonce": int(permit.meta.nonce),
        }

        domain = get_gasfree_domain(chain_id, controller)

        self._logger.info(f"Settling GasFree via Official HTTP API Proxy...")

        try:
            tx_hash = await api_client.submit(domain=domain, message=message, signature=signature)
            return tx_hash
        except Exception as e:
            self._logger.error(f"GasFree API submission failed: {e}")
            raise

File: typescript/packages/x402/src/mechanisms/gasfree.ts
import { TronGasFree } from '@gasfree/gasfree-sdk';
import {
  PaymentPayload,
  PaymentRequirements,
  PaymentPermit,
  DeliveryKind,
  ClientMechanism,
  ClientSigner,
  GASFREE_PRIMARY_TYPE,
  KIND_MAP,
  getChainId,
  getGasFreeApiBaseUrl,
  getGasFreeApiKey,
  getGasFreeApiSecret,
} from '../index.js';
import { GASFREE_TYPES, GasFreeAPIClient } from '../utils/gasfree.js';
import { findByAddress } from '../tokens.js';

export class GasFreeTronClientMechanism implements ClientMechanism {
  private signer: ClientSigner;

  constructor(signer: ClientSigner) {
    this.signer = signer;
  }

  scheme(): string {
    return 'gasfree_exact';
  }

  async createPaymentPayload(
    requirements: PaymentRequirements,
    resource: string,
    extensions?: Record<string, unknown>
  ): Promise<PaymentPayload> {
    const chainId = getChainId(requirements.network);
    const apiBaseUrl = getGasFreeApiBaseUrl(requirements.network);
    const apiKey = getGasFreeApiKey();
    const apiSecret = getGasFreeApiSecret();
    const userAddress = await this.signer.getAddress();
    
    const apiClient = new GasFreeAPIClient(apiBaseUrl, apiKey, apiSecret);

    // 1. Fetch account info
    console.debug(`Fetching account info for ${userAddress} from GasFree API...`);
    const accountInfo = await apiClient.getAddressInfo(userAddress);
    const gasfreeAddress = accountInfo.gasFreeAddress;
    
    if (!gasfreeAddress) {
      throw new Error(`Could not retrieve GasFree address for ${userAddress}`);
    }

    // 2. Check activation status
    const allowSubmit = (accountInfo as any).allowSubmit || false;
    if (!accountInfo.active && !allowSubmit) {
      throw new Error(`GasFree account for ${userAddress} (${gasfreeAddress}) is not activated.`);
    }

    // 3. Prepare maxFee and providers
    const providers = await apiClient.getProviders();
    if (!providers || providers.length === 0) {
        throw new Error('No GasFree service providers available');
    }
    const serviceProviderAddr = providers[Math.floor(Math.random() * providers.length)].address;
    console.debug(`Selected GasFree provider: ${serviceProviderAddr}`);

    const asset = accountInfo.assets.find((a: any) => a.tokenAddress === requirements.asset);
    const transferFee = BigInt(asset?.transferFee || '0');
    
    let maxFee = requirements.extra?.fee?.feeAmount;
    if (!maxFee) {
        // Default to transferFee from API, or 1.0 token units as a safe fallback
        if (transferFee > 0n) {
            maxFee = transferFee.toString();
        } else {
            const tokenInfo = findByAddress(requirements.network, requirements.asset);
            const decimals = tokenInfo?.decimals ?? 6;
            maxFee = (10 ** decimals).toString();
        }
    }

    let maxFeeBig = BigInt(maxFee);
    if (maxFeeBig < transferFee) {
      console.debug(`Increasing maxFee to ${transferFee} to meet protocol requirement`);
      maxFeeBig = transferFee;
      maxFee = maxFeeBig.toString();
    }

    // 4. Balance verification
    const skipBalanceCheck = (extensions as any)?.skipBalanceCheck || false;
    if (!skipBalanceCheck) {
      if (asset) {
          const assetBalance = BigInt(asset.balance || '0');
          const requiredTotal = BigInt(requirements.amount) + maxFeeBig;
          if (assetBalance < requiredTotal) {
            throw new Error(`Insufficient balance in GasFree wallet ${gasfreeAddress}.`);
          }
      } else {
        throw new Error(`Asset ${requirements.asset} not found in GasFree account ${gasfreeAddress}.`);
      }
    }

    const deadline = (extensions as any)?.paymentPermitContext?.meta?.validBefore || Math.floor(Date.now() / 1000) + 3600;
    
    const gasFree = new TronGasFree({ chainId });

    // 5. Sign TIP-712 typed data
    const { domain, types, message } = gasFree.assembleGasFreeTransactionJson({
      token: requirements.asset,
      serviceProvider: serviceProviderAddr,
      user: userAddress,
      receiver: requirements.payTo,
      value: requirements.amount,
      maxFee: maxFee,
      deadline: deadline.toString(),
      version: '1',
      nonce: accountInfo.nonce.toString(),
    });
    
    const signature = await this.signer.signTypedData(domain, types, message, GASFREE_PRIMARY_TYPE);

    // 6. Build PaymentPermit structure
    const paymentPermit: PaymentPermit = {
      meta: {
        kind: 'PAYMENT_ONLY' as DeliveryKind,
        paymentId: (extensions as any)?.paymentPermitContext?.meta?.paymentId || '',
        nonce: accountInfo.nonce.toString(),
        validAfter: (extensions as any)?.paymentPermitContext?.meta?.validAfter || 0,
        validBefore: Number(deadline),
      },
      buyer: userAddress,
      caller: serviceProviderAddr, 
      payment: {
        payToken: requirements.asset,
        payAmount: requirements.amount,
        payTo: requirements.payTo,
      },
      fee: {
        feeTo: serviceProviderAddr,
        feeAmount: maxFee,
      },
    };

    return {
      x402Version: 2,
      resource: { url: resource },
      accepted: requirements,
      payload: {
        signature,
        paymentPermit,
      },
      extensions: {
        ...extensions,
        gasfreeAddress,
        scheme: 'gasfree_exact',
      },
    };
  }
}
